worker/transfer.go

```diff
@@ -274,65 +273,73 @@ func uploadSlab(ctx context.Context, sp storeProvider, r io.Reader, m, n uint8,
 	return s, length, slowHosts, nil
 }
 
-func parallelDownloadSlab(ctx context.Context, sp storeProvider, ss object.SlabSlice, contracts []api.ContractMetadata, downloadSectorTimeout time.Duration, logger *zap.SugaredLogger) ([][]byte, []int64, error) {
-	// prepopulate the timings with a value for all contracts to ensure unused hosts aren't necessarily favoured in consecutive downloads
-	timings := make([]int64, len(contracts))
-	for i := 0; i < len(contracts); i++ {
+func parallelDownloadSlab(ctx context.Context, sp storeProvider, ss object.SlabSlice, hosts []api.ContractMetadata, downloadSectorTimeout time.Duration, logger *zap.SugaredLogger) ([][]byte, []int64, error) {
+	// prepopulate the timings with a value for all hosts to ensure unused hosts aren't necessarily favoured in consecutive downloads
+	timings := make([]int64, len(hosts))
+	for i := 0; i < len(hosts); i++ {
 		timings[i] = int64(defaultSectorDownloadTiming)
 	}
 
 	// ensure the context is cancelled when the slab is downloaded
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	// check whether we can recover the slab
-	if len(contracts) < int(ss.MinShards) {
-		return nil, nil, errors.New("not enough hosts to recover slab")
+	// make sure that there are enough hosts for the download.
+	hostMap := make(map[types.PublicKey]struct{})
+	for _, h := range hosts {
+		_, duplicate := hostMap[h.HostKey]
+		if duplicate {
+			panic("passed in hosts contains a duplicate")
```

Comment:
`panic` feels terribly aggressive? we could deduplicate and log a (serious) warning? Not allowing a download to succeed is a bit harsh especially in the context of migrations?